stan/math/opencl/cholesky_decompose.hpp

```diff
@@ -11,40 +11,38 @@
 #include <stan/math/opencl/subtract.hpp>
 #include <stan/math/opencl/err/check_diagonal_zeros.hpp>
 #include <stan/math/opencl/err/check_nan.hpp>
+#include <stan/math/opencl/err/check_opencl.hpp>
+#include <stan/math/opencl/sub_block.hpp>
 #include <CL/cl.hpp>
 #include <algorithm>
 #include <cmath>
 
 namespace stan {
 namespace math {
 /**
- * Return the lower-triangular Cholesky factor (i.e., matrix
- * square root) of the specified square, symmetric matrix.
- * The return value \f$L\f$ will be a lower-traingular matrix such that the
- * original matrix \f$A\f$ is given by
- * <p>\f$A = L \times L^T\f$.
- * The Cholesky decomposition is computed using an OpenCL kernel. This algorithm
- * is recursive. The matrix is subset into a matrix of size
- *  <code>A.rows() / 4</code>, and if the submatrix size is less than
- * 50 or <code>min_block</code> then the cholesky decomposition on the OpenCL
- * device is computed using that submatrix. If the submatrix is greater than
- * 50 or <code>min_block</code> then <code>cholesky_decompose</code> is run
- * again on a submatrix with size equal to <code>submat.rows() / 4</code>.
- * Once the Cholesky Decomposition is computed, the full matrix cholesky
- * is created by propogating the cholesky forward as given in the reference
- * report below.
+ * Performs an in-place of the the lower-triangular Cholesky factor (i.e.,
+ * matrix square root) of the specified square, symmetric matrix. The return
+ * value \f$L\f$ will be a lower-traingular matrix such that the original matrix
+ * \f$A\f$ is given by <p>\f$A = L \times L^T\f$. The Cholesky decomposition is
+ * computed using an OpenCL kernel. This algorithm is recursive. The matrix is
+ * subset into a matrix of size <code>A.rows() / 4</code>, and if the submatrix
+ * size is less than 50 or <code>min_block</code> then the cholesky
+ * decomposition on the OpenCL device is computed using that submatrix. If the
+ * submatrix is greater than 50 or <code>min_block</code> then
+ * <code>cholesky_decompose</code> is run again on a submatrix with size equal
+ * to <code>submat.rows() / 4</code>. Once the Cholesky Decomposition is
+ * computed, the full matrix cholesky is created by propogating the cholesky
+ * forward as given in the reference report below.
  *
  * For a full guide to how this works
  * see the Cholesy decompostion chapter in the  reference report
  * <a href="https://goo.gl/6kWkJ5"> here</a>.
- * @param A Symmetric matrix on the OpenCL device.
- * @return Square root of matrix on the OpenCL device.
  * @throw std::domain_error if m is not
  *  positive definite (if m has more than 0 elements)
  */
-inline matrix_cl cholesky_decompose(matrix_cl& A) {
+inline void cholesky_decompose(matrix_cl&& A) {
```

Comment:
Is this going to be annoying to have to do remember to do a move here or is this good? We could also just make the return type a `matrix_cl` (with the signature being const and making the buffer mutable?). I've been trying to read about how RVO works and whether a copy would be made here

Reply:
Check out https://youtu.be/xnqTKD8uD64?t=3068, really helped clarify my thinking on this stuff. There's even a handy chart (but I recommend listening to the 15 minutes or whatever of his talk on this subject).

![default C++ function parameter advice herb](https://user-images.githubusercontent.com/805356/56363918-496a6980-61bb-11e9-8bca-f44db526cbea.png)

Reply:
cool